### PR TITLE
[trainer, ckpt] fix: skip dataloader state restore when resuming at epoch boundary

### LINKED ISSUE #7401: [Bug][SFT] Resuming from an epoch-boundary checkpoint silently skips the next epoch
## Severity

High. A resumed multi-epoch SFT job can exit with status 0 without performing the requested remaining updates. The failure is silent, so an operator or scheduler can treat an under-trained model as a completed run.

## System Info

- Reproduced on latest verl main commit: `09ac37258ea66b0cb69b2738eec3074ea4e7261c`
- Linux 5.15 x86_64
- Python 3.12.13
- PyTorch 2.11.0+cu128, CUDA 12.8
- torchdata 0.11.0
- Ray 2.56.1
- Transformers 5.10.4
- GPU validation: 4 × NVIDIA RTX 3090 24 GiB, driver 580.173.02
- Model: Qwen3.5-0.8B (852.99M parameters)
- Ray SFT, FSDP2 size 4, BF16 mixed precision, synchronous checkpoint save

## Information

- [x] The official SFT trainer and shared checkpoint handler
- [x] My own minimal reproduction and four-GPU validation harness

## Tasks

- [x] The officially supported SFT training path
- [x] A synthetic 64-sample SFT dataset used only for the GPU A/B

## Reproduction

Save the reproducer below as `reproduce.py`. It uses the production `CheckpointHandler` and
`StatefulDataLoader`; only the model checkpoint engine is replaced by a minimal directory stub so
that the dataloader contract can be tested on CPU.

<details>
<summary>reproduce.py</summary>

```python
import argparse
import os
import tempfile

import torch
from torch.utils.data import DistributedSampler, TensorDataset
from torchdata.stateful_dataloader import StatefulDataLoader

from verl.utils.checkpoint import CheckpointHandler, OrchestrationMode


class Engine:
    def save_checkpoint(self, local_path, global_step=0, max_ckpt_to_keep=None, **kwargs):
        os.makedirs(local_path, exist_ok=True)

    def load_checkpoint(self, local_path, **kwargs):
        assert os.path.isdir(local_path)

    def is_mp_src_rank_with_outputs(self):
        return True

    def get_data_parallel_rank(self):
        return 0


def make_loader():
    dataset = TensorDataset(torch.arange(8))
    sampler = DistributedSampler(
        dataset, num_replicas=1, rank=0, shuffle=True, seed=0, drop_last=True
    )
    loader = StatefulDataLoader(
        dataset, batch_size=2, sampler=sampler, num_workers=0, drop_last=True
    )
    return sampler, loader


def batch_values(loader):
    return [batch[0].tolist() for batch in loader]


def reproduce_once(root):
    _, loader = make_loader()
    iterator = iter(loader)
    consumed = [next(iterator) for _ in range(len(loader))]
    CheckpointHandler(
        engine=Engine(), train_dataloader=loader, default_local_dir=root,
        resume_mode="disable", mode=OrchestrationMode.RAY,
    ).save_checkpoint(step=len(loader))

    sampler, resumed_loader = make_loader()
    resume_step = CheckpointHandler(
        engine=Engine(), train_dataloader=resumed_loader, default_local_dir=root,
        resume_mode="auto", mode=OrchestrationMode.RAY,
    ).load_checkpoint()
    sampler.set_epoch(resume_step // len(resumed_loader))
    return resume_step, consumed, batch_values(resumed_loader)


parser = argparse.ArgumentParser()
parser.add_argument("--expect", choices=("bug", "fixed"), required=True)
args = parser.parse_args()
for run in range(1, 4):
    with tempfile.TemporaryDirectory() as root:
        step, consumed, resumed = reproduce_once(root)
        print(
            f"run={run} resume_step={step} consumed_batches={len(consumed)} "
            f"resumed_batches={len(resumed)} resumed={resumed}"
        )
        assert resumed == ([] if args.expect == "bug" else [[5, 4], [2, 6], [7, 3], [1, 0]])
```

</details>

Steps:

1. Create an eight-sample dataset with batch size 2, so one epoch has four batches.
2. Consume all four batches and save `global_step_4` before requesting the iterator's next `StopIteration`, matching SFT's checkpoint timing.
3. Create a fresh loader and handler, resume step 4, set the sampler to epoch 1, and iterate.
4. Repeat three times.

```bash
(cd /tmp && \
  PYTHONNOUSERSITE=1 \
  PYTHONPATH=/path/to/verl \
  /path/to/python /path/to/reproduce.py --expect bug)
```

All three runs on main produce:

```text
run=1 resume_step=4 consumed_batches=4 resumed_batches=0 resumed=[]
run=2 resume_step=4 consumed_batches=4 resumed_batches=0 resumed=[]
run=3 resume_step=4 consumed_batches=4 resumed_batches=0 resumed=[]
```

A fresh epoch-1 loader yields four batches. The resumed loader should do the same.

The real GPU reproduction used Qwen3.5-0.8B, 64 selected samples, global batch size 16, four FSDP2 ranks, two epochs, and `checkpoint.async_save=false`. Baseline and candidate resumed byte-identical copies of the same complete step-4 checkpoint.

| Path | Exit code | Training steps after resume | Final tracker |
|---|---:|---|---:|
| main `09ac3725` | 0 | none | 4 |
| proposed fix `e06deb32` | 0 | 5, 6, 7, 8 | 8 |

The baseline log confirms that `data_0.pt` was loaded, reports the automatic resume from global step 4, enters `Epoch 2/2`, then yields zero batches and exits normally.

## Expected behavior

An epoch-boundary checkpoint represents a completed saved epoch. Resuming should leave the new loader fresh for the next epoch and execute steps 5–8.

## Actual behavior

SFT restores the exhausted saved loader state at the beginning of the next epoch. The first iteration yields no batches, the outer epoch loop ends, and the process exits with status 0. The tracker stays at step 4.

## Root cause

`CheckpointHandler.load_checkpoint()` unconditionally restores `data_<dp_rank>.pt`. At an epoch boundary that state contains no remaining samples. Both SFT trainers then calculate the next epoch with:

```python
start_epoch = global_step // steps_per_epoch
```

The exhausted state is therefore applied to the next epoch rather than discarded.

## Impact

- A preempted or manually resumed multi-epoch SFT run can silently omit an entire remaining epoch.
- No exception or nonzero exit code indicates that the configured training target was missed.
- The shared handler is used by both local SPMD and Ray SFT. The end-to-end GPU evidence above specifically validates Ray/FSDP2 with synchronous checkpoint saving.

## Proposed fix

- Save versioned epoch-position metadata beside every SFT dataloader state: checkpoint step, batches per epoch, and step within the epoch.
- Validate the metadata against the selected checkpoint and current number of batches per epoch.
- Skip dataloader restoration only at a confirmed epoch boundary; continue restoring mid-epoch state normally.
- Keep old checkpoints compatible by inferring their boundary from the current loader length and warning about that limitation.
- Return without training if the resumed step already reached `total_training_steps`.
- In SPMD mode, publish the tracker and optional HDFS copy only after all dataloader files have been written.

The metadata detects a changed number of batches per epoch; it does not claim to identify arbitrary dataset, sampler, or DP-topology changes that happen to preserve that number. This proposal also does not claim full directory-level crash atomicity or cover Megatron asynchronous checkpoint finalization.

## Validation of the proposed fix

- CPU production-handler reproduction: main failed 3/3; the fix yielded all four next-epoch batches in 3/3.
- Focused tests: 14 passed.
- Focused plus related checkpoint tests: 46 passed.
- All 13 configured pre-commit hooks passed.
- Four-GPU formal A/B: all four phases exited 0.
- The fixed boundary resume ran steps 5–8 and advanced the tracker to 8.
- The fixed mid-epoch control resumed step 2 and ran only steps 3–4; token counts 1096 and 1080 exactly matched the uninterrupted run's steps 3–4.
- Final step-4 and step-8 checkpoints each contained model, optimizer, and extra-state shards for ranks 0–3 plus dataloader state. The frozen formal validator passed ZIP CRC for all 13 step-8 `.pt` files. A separate independent read-only audit also checked and passed all 13 step-4 files.
- No traceback, RuntimeError, OOM, NCCL failure, worker crash, NaN, or Inf appeared in the four authoritative logs.

## Duplicate search

All-state Issue and PR searches on 2026-08-14 found no exact SFT implementation:

- [Issues: `"StatefulDataLoader" "epoch boundary" SFT`](https://github.com/verl-project/verl/issues?q=is%3Aissue%20%22StatefulDataLoader%22%20%22epoch%20boundary%22%20SFT)
- [PRs: `"StatefulDataLoader" "epoch boundary" SFT`](https://github.com/verl-project/verl/pulls?q=is%3Apr%20%22StatefulDataLoader%22%20%22epoch%20boundary%22%20SFT)
- [PRs: `"CheckpointHandler" "epoch boundary"`](https://github.com/verl-project/verl/pulls?q=is%3Apr%20%22CheckpointHandler%22%20%22epoch%20boundary%22)

Merged PR [#5725](https://github.com/verl-project/verl/pull/5725) fixes the same exhausted-state mechanism only in `verl/trainer/ppo/ray_trainer.py`. SFT restores through `verl/utils/checkpoint/checkpoint_handler.py`, where the unconditional load remains on main. Generic SFT resume Issue [#609](https://github.com/verl-project/verl/issues/609) and merged resume support PR [#2292](https://github.com/verl-project/verl/pull/2292) do not report or handle this silent epoch-boundary failure.

### PR DESCRIPTION
**Pull request by ewan0x79**
### What does this PR do?

Fixes #7401. When a checkpoint is saved exactly at an epoch boundary (e.g. `save_freq` is a multiple of `steps_per_epoch`), resuming silently skips the next epoch: the job can exit with status 0 having performed none of the remaining updates, leaving an under-trained model that looks "complete".

**Root cause:** the saved `StatefulDataLoader` state marks the sampler as fully exhausted. On resume, the trainer already advances to the next epoch (`start_epoch = global_step // steps_per_epoch`) and calls `set_epoch()` before iterating; the restored iterator then fast-forwards the *new* epoch's sampler by the whole epoch length (torchdata restores the sampler position via `itertools.islice`), so the new epoch yields 0 batches.

**Fix:** #5725 solved the same bug for `ppo/ray_trainer.py` with a modulo check (`global_steps % steps_per_epoch == 0`). This PR applies the same, already-accepted approach to the remaining affected paths:

- `verl/utils/checkpoint/checkpoint_handler.py` (shared by SFT spmd/ray trainers): new optional `steps_per_epoch` argument; skip restoring the dataloader state when `resume_global_step % steps_per_epoch == 0`. Detection is disabled when the argument is not provided, so existing behavior is unchanged for other callers.
- `verl/trainer/sft_trainer.py` / `verl/trainer/sft_trainer_ray.py`: pass `steps_per_epoch` through.
- `verl/trainer/ppo/v1/trainer_base.py`: same boundary check as `ppo/ray_trainer.py`, using the trainer's own `steps_per_epoch` (consistent with its `current_epoch` computation).

**Relation to #7402 (why this is not a duplicate):** #7402 addresses the same issue but introduces a new on-disk sidecar metadata file (~50 lines plus format versioning and a legacy path), which only protects *newly written* checkpoints and adds a new hard failure (`ValueError`) when resuming with a different `steps_per_epoch` (e.g. different GPU count after preemption). This PR is the minimal fix: no new artifact, works for every existing checkpoint, adds no new failure mode, and also covers `ppo/v1/trainer_base.py`, which #7402 does not touch. Related: #5725, #7402, #7516.

### Checklist Before Starting

- [x] Search for similar PRs. Paste at least one query link here: https://github.com/verl-project/verl/pulls?q=is%3Apr+dataloader+epoch+boundary
- [x] Format the PR title as `[{modules}] {type}: {description}` (This will be checked by the CI)

### Test

Added `tests/utils/ckpt/test_checkpoint_handler_epoch_boundary_on_cpu.py` (9 cases, matches the `tests/**/test_*_on_cpu.py` pattern so it is picked up by `cpu_unit_tests.yml` automatically):

```bash
python -m pytest tests/utils/ckpt/test_checkpoint_handler_epoch_boundary_on_cpu.py -q
# 9 passed
```

Covers: boundary detection on/off/mid-epoch/first-epoch steps, and `_load_dataloader_state` end-to-end (stale state dropped at boundary / restored mid-epoch / restored when detection disabled / missing file starts from scratch).

Also validated end-to-end with a real `StatefulDataLoader` (torchdata 0.11.0, `num_workers=2`) reproducing the verl loop (`set_epoch` + `iter` per epoch, saving inside the loop body before `StopIteration`): without the fix the resumed epoch yields **0 batches**; with the fix it yields the full epoch.

Observed in production: a 4-epoch Megatron SFT job preempted at global step 4225 resumed from `global_step_3183` (`3 × steps_per_epoch`) and exited SUCCEEDED after 0 training steps; only 3 epoch checkpoints existed while 4 were configured.

### API and Usage Example

No API changes. `CheckpointHandler` gains an optional keyword argument `steps_per_epoch` (default `None` = detection off, fully backward compatible). No user action is required.

```python
# SFT trainers already wire this up internally; no user-side change.
# For direct CheckpointHandler users who want the protection:
handler = CheckpointHandler(
    engine=engine,
    train_dataloader=train_dataloader,
    default_local_dir=ckpt_dir,
    steps_per_epoch=len(train_dataloader),  # optional; enables epoch-boundary detection
)
```

### Design & Code Changes

- `verl/utils/checkpoint/checkpoint_handler.py`: optional `steps_per_epoch`; `_at_epoch_boundary()` helper; `_load_dataloader_state()` logs a warning and skips the stale restore at epoch boundaries.
- `verl/trainer/sft_trainer.py`, `verl/trainer/sft_trainer_ray.py`: pass `steps_per_epoch` (both call sites are after `steps_per_epoch` is computed).
- `verl/trainer/ppo/v1/trainer_base.py`: same boundary check as `ppo/ray_trainer.py` (#5725), based on the trainer's own `steps_per_epoch`.
- `tests/utils/ckpt/test_checkpoint_handler_epoch_boundary_on_cpu.py`: new unit tests (9 cases).

### Checklist Before Submitting

- [x] Read the [Contribute Guide](https://github.com/verl-project/verl/blob/main/CONTRIBUTING.md).
- [x] Apply [pre-commit checks](https://github.com/verl-project/verl/blob/main/CONTRIBUTING.md#code-linting-and-formatting): all hooks pass (ruff, ruff-format, mypy, autogen-trainer-cfg, license, docstrings, etc.).
- [x] Add / Update [the documentation](https://github.com/verl-project/verl/tree/main/docs): not applicable — bug fix restoring expected behavior, no user-facing API or config change.
- [x] Add unit or end-to-end test(s) to [the CI workflow](https://github.com/verl-project/verl/tree/main/.github/workflows) to cover all the code: new CPU unit tests auto-discovered by `cpu_unit_tests.yml` (`tests/**/test_*_on_cpu.py`).
- [ ] Once your PR is ready for CI, send a message in [the `ci-request` channel](https://verl-project.slack.com/archives/C091TCESWB1) in [the `verl` Slack workspace](https://join.slack.com/t/verl-project/shared_invite/zt-3855yhg8g-CTkqXu~hKojPCmo7k_yXTQ).
- [x] Recipe submodule: not related.

---

**AI-assistance disclosure (per AGENTS.md):** this PR was prepared with AI assistance (Claude). I have reviewed every changed line and run the tests listed above. Why this is not a duplicate: see "Relation to #7402" — materially different approach (minimal modulo check as already merged in #5725 vs. a new on-disk sidecar format), plus coverage of `ppo/v1/trainer_base.py`. Test commands and results: listed in the Test section (9/9 passed, plus the end-to-end before/after comparison).

### CODE REVIEW

**Review by koriyoshi2041 — APPROVED** [review]
Reviewed exact head `4105a991`. The boundary predicate is consistent across all three affected resume paths: both SFT trainers pass `len(train_dataloader)` into the shared handler before `load_checkpoint()`, and v1 uses its existing global-step epoch denominator. The sibling path remains intact: step 0 and mid-epoch checkpoints still restore the saved loader state, while only positive exact multiples discard the exhausted iterator. This also matches the already-merged PPO precedent in #5725 without adding #7402’s sidecar format or topology-sensitive hard failure. I checked the complete five-file delta and `git diff --check` passes. The new 9-case CPU suite pins disabled, scratch, mid-epoch, boundary, first-boundary, missing-file, and actual state-load behavior.
